### PR TITLE
Adds related demos in sidebar 

### PR DESCRIPTION
**Pull request by glassnotes**
**Title:** Adds related demos in sidebar

**Summary:** This PR adds a new component in the sidebar called "Related demos" so that we can link related tutorials together.  

**Relevant references:** None

**Possible Drawbacks:** None

**Related GitHub Issues:** None

### CODE REVIEW

**Comment by glassnotes** [comment]
The demos have been updated with links to each other - below is a list of less well-connected ones. Let me know here if there are any suggestions or relationships I've missed.

Don't link to anything:
 - Basic tutorial: qubit rotation
 - Gaussian transformation
 - Optimizing noisy circuits with Cirq
 - Ensemble classification with Forest and Qiskit

Nothing links to them:
 - 3-qubit Ising model with PyTorch
 - Plugins and hybrid computation
 - QAOA maxcut (will soon be linked with Intro QAOA tutorial)
 - Quantum chemistry with PennyLane
 - Variationally optimizing measurement protocols
 - Frugal shot optimization with Rosalin
 - Quantum circuit structure learning
   
Orphans:
 - QGRNN
 - QGANs
 - Quantum models as Fourier series
 - Quanvolutional NNs
 - Quantum transfer learning
 - Quantum computation with neutral atoms

File: demonstrations/pytorch_noise.py [hunk 1]
```diff
@@ -9,6 +9,10 @@
         by using it to optimize a noisy quantum hardware device.
     :property="og:image": https://pennylane.ai/qml/_images/bloch.gif
 
+.. related::
+
+   tutorial_noisy_circuit_optimization Optimizing noisy circuits with Cirq
```
Comment (josh146) [inline demonstrations/pytorch_noise.py:14]: Sidenote: it's a shame the title can't be automatically extracted from the filename. I think I spent a while trying to do this, and couldn't get it to work.

Comment (josh146) [reply demonstrations/pytorch_noise.py:14]: I think it's possible, just requires some advanced knowledge of how `docutils` works, in order to extract the titles from the parser inside the `RelatedDirective` class :)

Comment (josh146) [inline demonstrations/pytorch_noise.py:14]: How come the related links aren't bi-directional? For example, I would expect "Optimizing noisy circuits with Cirq" to link back to here?

File: demonstrations/tutorial_backprop.py [hunk 2]
```diff
@@ -7,6 +7,11 @@
 
     :property="og:image": https://pennylane.ai/qml/_images/sphx_glr_tutorial_backprop_002.png
 
+.. related::
+
+   parameter_shift Parameter-shift rules
+   tutorial_quantum_natural_gradient Quantum natural gradient
```
Comment (josh146) [inline demonstrations/tutorial_backprop.py:13]: This is great!

File: demonstrations/tutorial_data_reuploading_classifier.py [hunk 3]
```diff
@@ -1,14 +1,19 @@
 r"""
 .. _data_reuploading_classifier:
 
-Data-reuploading classifer
-==========================
+Data-reuploading classifier
```
Comment (josh146) [inline demonstrations/tutorial_data_reuploading_classifier.py:4]: Oops. Good catch. This is one of our most popular pages as well

File: demonstrations/tutorial_plugins_hybrid.py [hunk 4]
```diff
@@ -10,6 +10,11 @@
         containing both photonic and qubit devices.
     :property="og:image": https://pennylane.ai/qml/_images/photon_redirection.png
 
+.. related::
+
+   tutorial_qubit_rotation Basic tutorial: qubit rotation
```
Comment (josh146) [inline demonstrations/tutorial_plugins_hybrid.py:15]: Does it make sense to have Qubit Rotation also link to these tutorials? In this case, the 'Related tutorials' section acts as a sort of 'Where to go next' cue for the user.

I've had feedback before from users that sometimes they end up on qubit rotation (since we link to it in the 'Getting Started' section), and don't really know where to go next.

**Review by josh146 — APPROVED** [review]
thanks for doing this, it looks great @glassnotes 🌮 

This is an example of a relatively minor addition which I think leads to a significant navigation improvement for users!

**Comment by josh146** [comment]
That was meant to be a 🎉 emoji, but I'm leaving in the accidental 🌮

**Comment by mariaschuld** [comment]
Great, just two thoughts:

Frugal shot optimization with Rosalin
Quantum circuit structure learning
-> These could link to and get links from other "optimiser" tutorials like quantum natural gradient (as well as each other)?

Quantum transfer learning
-> When the embeddings tutorial is back live (currently under repair) we could link from it

File: demonstrations/tutorial_backprop.py [hunk 5]
```diff
@@ -7,6 +7,11 @@
 
     :property="og:image": https://pennylane.ai/qml/_images/sphx_glr_tutorial_backprop_002.png
 
+.. related::
+
+   parameter_shift Parameter-shift rules
```
Comment (mariaschuld) [inline demonstrations/tutorial_backprop.py:12]: I think this link is not working? At least one of the parameter-shift links gave me an "unknown artefact" error...maybe double check if you can reproduce.

**Review by mariaschuld — APPROVED** [review]
Very cool feature indeed...it's so much fun to click from one to the other!

File: demonstrations/tutorial_data_reuploading_classifier.py [hunk 6]
```diff
@@ -1,14 +1,19 @@
...
 *Author: Shahnawaz Ahmed (shahnawaz.ahmed95@gmail.com)*
 
 .. meta::
    :property="og:description": Implement a single-qubit universal quantum classifier using PennyLane.
    :property="og:image": https://pennylane.ai/qml/_images/universal_dnn1.png
 
+.. related::
+
+   tutorial_variational_classifier Variational quantum classifier
+   tutorial_multiclass_classification Multiclass margin classifier
```
Comment (mariaschuld) [inline demonstrations/tutorial_data_reuploading_classifier.py:15]: It's not obvious, but the Fourier series tutorial can be linked here (it explains some of the magic behind data re-uploading).

File: demonstrations/tutorial_doubly_stochastic.py [hunk 7]
```diff
@@ -7,6 +7,11 @@
         strategy with doubly stochastic gradient descent.
     :property="og:image": https://pennylane.ai/qml/_images/single_shot.png
 
+.. related::
+
+   tutorial_backprop Quantum gradients with backprop
+   tutorial_quantum_natural_gradient Quantum natural gradient
```
Comment (mariaschuld) [inline demonstrations/tutorial_doubly_stochastic.py:13]: I'd add the frugal shot optimiser tutorial (there is one, right?). The papers overlap quite a lot.

**Comment by glassnotes** [comment]
[ch1048]

File: demonstrations/tutorial_backprop.py (hunk 5 above)
Comment (glassnotes) [reply demonstrations/tutorial_backprop.py:12]: Yup, I thought it was a tutorial being referenced lower down, but it's actually just the glossary page so I'm removing this link.

File: demonstrations/tutorial_plugins_hybrid.py (hunk 4 above)
Comment (glassnotes) [reply demonstrations/tutorial_plugins_hybrid.py:15]: I've linked it to a few other basic tutorials now. But that brings up an interesting point, maybe we should try and design some tutorials that are "in between" the very basics, and applications/research-level stuff.

File: demonstrations/pytorch_noise.py (hunk 1 above)
Comment (glassnotes) [reply demonstrations/pytorch_noise.py:14]: I was trying to think from the standpoint of someone working sequentially through the tutorials, so a number of the links are one-directional. That said, there are probably multiple paths so maybe we should aim for bi-directionality.

Comment (co9olguy) [reply demonstrations/pytorch_noise.py:14]: I was thinking to have the bi-directional links to related tutorials, then somehow have a separate page/mechanism which curate a specific recommended curriculum & order?

File: demonstrations/tutorial_plugins_hybrid.py (hunk 4 above)
Comment (co9olguy) [reply demonstrations/tutorial_plugins_hybrid.py:15]: Absolutely!

**Comment by co9olguy** [comment]
This is awesome to have, thanks @glassnotes!

I wonder if the sidebar is a bit hidden? e.g., maybe in the future we could think about having 2-3 related tutorials at the top or bottom in more noticeable locations?

File: demonstrations/tutorial_isingmodel_PyTorch.py [hunk 8]
```diff
@@ -9,6 +9,11 @@
         PennyLane to optimize a 3-qubit Ising model.
     :property="og:image": https://pennylane.ai/qml/_images/isingspins.png
 
+.. related::
+
+   tutorial_state_preparation Training a quantum circuit with PyTorch
+   pytorch_noise PyTorch noise
```
Comment (co9olguy) [inline demonstrations/tutorial_isingmodel_PyTorch.py:15]: Note that the "caption" for this tutorial is different in different places. Is there a way to automatically ensure that the title is the same no matter which tutorial it appears in?

Comment (co9olguy) [reply demonstrations/tutorial_isingmodel_PyTorch.py:15]: e.g., elsewhere it's called "Pytorch and noisy devices"

**Comment by co9olguy** [comment]
Just did a fly-by, won't review in detail, but this is a great addition

File: demonstrations/tutorial_isingmodel_PyTorch.py (hunk 8 above)
Comment (glassnotes) [reply demonstrations/tutorial_isingmodel_PyTorch.py:15]: Oh good catch, that one is just my mistake :)

Comment (glassnotes) [reply demonstrations/tutorial_isingmodel_PyTorch.py:15]: (@josh146 thinks automating this would be possible, but challenging.)

**Comment by glassnotes** [comment]
> This is awesome to have, thanks @glassnotes!
> 
> I wonder if the sidebar is a bit hidden? e.g., maybe in the future we could think about having 2-3 related tutorials at the top or bottom in more noticeable locations?

I do like the idea of having them at the bottom because it shows you where to go next once you're done reading. 

On a related note, the bug with the footer eating the sidebar stuff sometimes hides the whole section for tutorials where the table of contents is very long. In principle the sidebar should be scrollable, I've noticed this doesn't become possible until you scroll all the way to the bottom of the page. So once that is fixed, they should be more visible.

File: demonstrations/pytorch_noise.py (hunk 1 above)
Comment (glassnotes) [reply demonstrations/pytorch_noise.py:14]: Yeah that would be awesome! I think I am going to save that for a future PR; my next planned work on the demos is focused more on the content, so that will be a better time to decide on an order (and figure out how to fill in the gaps).
